runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/SparkGroupAlsoByWindowViaWindowSet.java

```diff
@@ -73,440 +73,555 @@
+            return new Tuple2<>(encodedKey, new Tuple2<>(updated, serOutput));
+          }
+          // an empty state with no output, can be evicted completely - do nothing.
         }
+        return endOfData();
+      }
+    }
 
-        return scala.collection.JavaConversions.asScalaIterator(outIter);
+    private final FullWindowedValueCoder<InputT> wvCoder;
+    private final Coder<K> keyCoder;
+    private final List<Integer> sourceIds;
+    private final TimerInternals.TimerDataCoder timerDataCoder;
+    private final WindowingStrategy<?, W> windowingStrategy;
+    private final SerializablePipelineOptions options;
+    private final IterableCoder<WindowedValue<InputT>> itrWvCoder;
+    private final String logPrefix;
+    private final Coder<WindowedValue<KV<K, Iterable<InputT>>>> wvKvIterCoder;
+
+    UpdateStateByKeyFunction(
+        final List<Integer> sourceIds,
+        final WindowingStrategy<?, W> windowingStrategy,
+        final FullWindowedValueCoder<InputT> wvCoder,
+        final Coder<K> keyCoder,
+        final SerializablePipelineOptions options,
+        final String logPrefix) {
+      this.wvCoder = wvCoder;
+      this.keyCoder = keyCoder;
+      this.sourceIds = sourceIds;
+      this.timerDataCoder = timerDataCoderOf(windowingStrategy);
+      this.windowingStrategy = windowingStrategy;
+      this.options = options;
+      this.itrWvCoder = IterableCoder.of(wvCoder);
+      this.logPrefix = logPrefix;
+      this.wvKvIterCoder =
+          windowedValueKeyValueCoderOf(
+              keyCoder,
+              wvCoder.getValueCoder(),
+              ((FullWindowedValueCoder<InputT>) wvCoder).getWindowCoder());
+    }
+
+    @Override
+    public Iterator<
+            Tuple2</*K*/ ByteArray, Tuple2<StateAndTimers, /*WV<KV<K, Itr<I>>>*/ List<byte[]>>>>
+        apply(
+            final Iterator<
+                    Tuple3<
+                        /*K*/ ByteArray, Seq</*Itr<WV<I>>*/ byte[]>,
+                        Option<Tuple2<StateAndTimers, /*WV<KV<K, Itr<I>>>*/ List<byte[]>>>>>
+                input) {
+      //--- ACTUAL STATEFUL OPERATION:
+      //
+      // Input Iterator: the partition (~bundle) of a co-grouping of the input
+      // and the previous state (if exists).
+      //
+      // Output Iterator: the output key, and the updated state.
+      //
+      // possible input scenarios for (K, Seq, Option<S>):
+      // (1) Option<S>.isEmpty: new data with no previous state.
+      // (2) Seq.isEmpty: no new data, but evaluating previous state (timer-like behaviour).
+      // (3) Seq.nonEmpty && Option<S>.isDefined: new data with previous state.
+
+      final SystemReduceFn<K, InputT, Iterable<InputT>, Iterable<InputT>, W> reduceFn =
+          SystemReduceFn.buffering(wvCoder.getValueCoder());
+
+      // use in memory Aggregators since Spark Accumulators are not resilient
+      // in stateful operators, once done with this partition.
```

Comment:
`once done with this partition` maybe I was trying to say `log once done with this partition`?

Reply:
Yeah I think this comment can be removed, as aggregators are nowhere to be found, probably replaced by `MetricsContainer`s.